src/main/scala/org/graphframes/lib/Pregel.scala

```diff
@@ -0,0 +1,289 @@
+
+import org.graphframes.GraphFrame
+import org.graphframes.GraphFrame._
+import org.apache.spark.sql.{Column, DataFrame}
+import org.apache.spark.sql.functions.{array, col, explode, struct}
+
+/**
+ * Implements a Pregel-like bulk-synchronous message-passing API based on DataFrame operations.
+ *
+ * See <a href="https://doi.org/10.1145/1807167.1807184">Malewicz et al., Pregel: a system for large-scale graph
+ * processing</a> for a detailed description of the Pregel algorithm.
+ *
+ * You can construct a Pregel instance using either this constructor or [[org.graphframes.GraphFrame#pregel]],
+ * then use builder pattern to describe the operations, and then call [[run]] to start a run.
+ * It returns a DataFrame of vertices from the last iteration.
+ *
+ * When a run starts, it expands the vertices DataFrame using column expressions defined by [[withVertexColumn]].
+ * Those additional vertex properties can be changed during Pregel iterations.
+ * In each Pregel iteration, there are three phases:
+ *   - Given each triplet, generate messages and targets to send,
+ *     described by [[sendMsgToDst]] and [[sendMsgToSrc]].
+ *   - Aggregate messages by destination vertex IDs, described by [[aggMsgs]].
+ *
+ *   - Update additional vertex properties based on aggregated messages and states from previous iteration,
+ *     described by [[withVertexColumn]].
+ *
+ * Please find what columns you can reference at each phase in the method API docs.
+ *
+ * You can control the number of iterations by [[setMaxIter]] and check API docs for advanced controls.
+ *
+ * Example code for Page Rank:
+ *
+ * {{{
+ *   val edges = ...
+ *   val vertices = GraphFrames.fromEdges(edges).outDegrees.cache()
+ *   val numVertices = vertices.count()
+ *   val graph = GraphFrames(vertices, edges)
+ *   val alpha = 0.15
+ *   val ranks = graph.pregel
+ *     .withVertexColumn("rank", lit(1.0 / numVertices),
+ *       coalesce(Pregel.msg, lit(0.0)) * (1.0 - alpha) + alpha / numVertices)
+ *     .sendMsgToDst(Pregel.src("rank") / Pregel.src("outDegrees"))
+ *     .aggMsgs(sum(Pregel.msg))
+ *     .run()
+ * }}}
+ *
+ * @param graph The graph that Pregel will run on.
+ * @see [[org.graphframes.GraphFrame#pregel]]
+ * @see <a href="https://doi.org/10.1145/1807167.1807184">
+ *        Malewicz et al., Pregel: a system for large-scale graph processing.
+ *      </a>
+ */
+class Pregel(val graph: GraphFrame) {
+
+  private val withVertexColumnList = collection.mutable.ListBuffer.empty[(String, Column, Column)]
+
+  private var maxIter: Int = 10
+  private var checkpointInterval = 2
+
+  private var sendMsgs = collection.mutable.ListBuffer.empty[(Column, Column)]
+  private var aggMsgsCol: Column = null
+
+  private val CHECKPOINT_NAME_PREFIX = "pregel"
+
+  /** Sets the max number of iterations (default: 10). */
+  def setMaxIter(value: Int): this.type = {
+    maxIter = value
+    this
+  }
+
+  /**
+   * Sets the number of iterations between two checkpoints (default: 2).
```

Comment:
What about change default value to be `1` ? (In my test `localCheckpoint` will be fast and this change will probably save SQL planner time, especially when there're many partitions in dataset.)